Panic if can't find OS wordlist

On Arch Linux, guess I don't have an OS wordlist: 

`
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: ToipeError { msg: "No such file or directory (os error 2)" }', /home/ben/.cargo/registry/src/github.com-1ecc6299db9ec823/toipe-0.4.0/src/main.rs:9:40
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
`
Could print a warning or just use the default word list? I can write a fix for this since I want an excuse to do rust

If you look at my latest commit (benliepert/toipe@cd0951e6e33b39ea2c02fc9aa28cb855216faad2) you can see what I did to fix this. But, it's ugly as the match statement is repeated. Is this how you would fix it? can you see a way to consolidate the match statements?
I tried this:
```
pub fn new(config: ToipeConfig) -> Result<Self, ToipeError> {
        fn create_toipe_wordlist_error(e: std::io::Error) -> ToipeError {
            let mut toipe_error = ToipeError::from(e);
            toipe_error.msg = "Error reading the given word list: ".to_string() + &toipe_error.msg;
            toipe_error
        }
        let word_selector: Box<dyn WordSelector> = {
            let mut word_selector_result_option: Option<_>; //: Result<RawWordSelector<std::fs::File>, std::io::Error>;
            if let Some(wordlist_path) = config.wordlist_file.clone() {
                word_selector_result_option = Some(RawWordSelector::from_path(PathBuf::from(wordlist_path)));
            } else if let Some(word_list) = config.wordlist.contents() {
                word_selector_result_option = Some(RawWordSelector::from_string(word_list.to_string()));
            } else if let BuiltInWordlist::OS = config.wordlist {
                word_selector_result_option = Some(RawWordSelector::from_path(PathBuf::from(OS_WORDLIST_PATH)));
            } else {
                // this should never happen!
                // TODO: somehow enforce this at compile time?
                return Err(ToipeError {
                    msg: "Undefined word list or path.".to_string(),
                });
            };

            match word_selector_result_option {
                Some(thing) => {
                    match thing {
                        Ok(rws) => Box::new(rws),
                        Err(e) => return Err(create_toipe_wordlist_error(e)),
                    }
                }
                None => {
                    return Err(ToipeError {
                        msg: "This is logically impossible to occur".to_string(),
                    });
                }
            }
 
        };
```
but I get this compilation error:
```
error[E0308]: mismatched types
  --> src/lib.rs:79:52
   |
79 |                 word_selector_result_option = Some(RawWordSelector::from_string(word_list.to_string()));
   |                                                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected struct `File`, found struct `std::io::Cursor`
   |
   = note: expected enum `Result<RawWordSelector<File>, _>`
              found enum `Result<RawWordSelector<std::io::Cursor<std::string::String>>, _>`

For more information about this error, try `rustc --explain E0308`.
error: could not compile `toipe` due to previous error
```
it's not making sense to me because it looks like from_string() and from_path() both return the same type.

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -71,7 +71,18 @@ impl<'a> Toipe {
             } else if let Some(word_list) = config.wordlist.contents() {
                 Box::new(RawWordSelector::from_string(word_list.to_string())?)
             } else if let BuiltInWordlist::OS = config.wordlist {
-                Box::new(RawWordSelector::from_path(PathBuf::from(OS_WORDLIST_PATH))?)
+                //Box::new(RawWordSelector::from_path(PathBuf::from(OS_WORDLIST_PATH))?)
+                let rws_result = RawWordSelector::from_path(PathBuf::from(OS_WORDLIST_PATH));
+                match rws_result {
+                    Ok(rws) => Box::new(rws),
+                    Err(e) => {
+                        // convert to toipe error
+                        let mut te = ToipeError::from(e);
+                        let msg_addition = "Error reading the given word list: ".to_string();
+                        te.msg = msg_addition + &te.msg;
+                        return Err(te);
+                    }
+                }
             } else {
                 // this should never happen!
                 // TODO: somehow enforce this at compile time?

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -43,6 +43,16 @@ pub struct ToipeError {
     pub msg: String,
 }
 
+impl ToipeError {
+    /// Prefixes a ToipeError's msg with 'msg_context'
+    ///
+    /// The ToipeError operated on is consumed, and a new one is returned
+    pub fn with_context(mut self: Self, msg_context: &str) -> Self {
+        self.msg = msg_context.to_owned() + &self.msg;
+        self
+    }
+}
+
 /// Converts [`std::io::Error`] to [`ToipeError`].
 ///
 /// This keeps only the error message.
@@ -65,30 +75,25 @@ impl<'a> Toipe {
     /// Initializes the word selector.
     /// Also invokes [`Toipe::restart()`].
     pub fn new(config: ToipeConfig) -> Result<Self, ToipeError> {
-        fn create_toipe_wordlist_error(e: std::io::Error) -> ToipeError {
-            let mut toipe_error = ToipeError::from(e);
-            toipe_error.msg = "Error reading the given word list: ".to_string() + &toipe_error.msg;
-            toipe_error
-        }
-
+        let word_selector_error_prefix = "Error reading the given word list: ";
         let word_selector: Box<dyn WordSelector> =
             if let Some(wordlist_path) = config.wordlist_file.clone() {
                 let word_selector_result = RawWordSelector::from_path(PathBuf::from(wordlist_path));
                 match word_selector_result {
                     Ok(rws) => Box::new(rws),
-                    Err(e) => return Err(create_toipe_wordlist_error(e)),
+                    Err(e) => return Err(ToipeError::from(e).with_context(word_selector_error_prefix)),
                 }
             } else if let Some(word_list) = config.wordlist.contents() {
                 let word_selector_result = RawWordSelector::from_string(word_list.to_string());
                 match word_selector_result {
                     Ok(rws) => Box::new(rws),
-                    Err(e) => return Err(create_toipe_wordlist_error(e)),
+                    Err(e) => return Err(ToipeError::from(e).with_context(word_selector_error_prefix)),
                 }
             } else if let BuiltInWordlist::OS = config.wordlist {
                 let word_selector_result = RawWordSelector::from_path(PathBuf::from(OS_WORDLIST_PATH));
                 match word_selector_result {
                     Ok(rws) => Box::new(rws),
-                    Err(e) => return Err(create_toipe_wordlist_error(e)),
+                    Err(e) => return Err(ToipeError::from(e).with_context(word_selector_error_prefix)),
                 }
             } else {
                 // this should never happen!

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -44,11 +44,9 @@ pub struct ToipeError {
 }
 
 impl ToipeError {
-    /// Prefixes a ToipeError's msg with 'msg_context'
-    ///
-    /// The ToipeError operated on is consumed, and a new one is returned
-    pub fn with_context(mut self: Self, msg_context: &str) -> Self {
-        self.msg = msg_context.to_owned() + &self.msg;
+    /// Prefixes the message with a context
+    pub fn with_context(mut self: Self, context: &str) -> Self {
+        self.msg = context.to_owned() + &self.msg;
         self
     }
 }

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -65,23 +65,30 @@ impl<'a> Toipe {
     /// Initializes the word selector.
     /// Also invokes [`Toipe::restart()`].
     pub fn new(config: ToipeConfig) -> Result<Self, ToipeError> {
+        fn create_toipe_wordlist_error(e: std::io::Error) -> ToipeError {
+            let mut toipe_error = ToipeError::from(e);
+            toipe_error.msg = "Error reading the given word list: ".to_string() + &toipe_error.msg;
+            toipe_error
+        }
+
         let word_selector: Box<dyn WordSelector> =
             if let Some(wordlist_path) = config.wordlist_file.clone() {
-                Box::new(RawWordSelector::from_path(PathBuf::from(wordlist_path))?)
-            } else if let Some(word_list) = config.wordlist.contents() {
-                Box::new(RawWordSelector::from_string(word_list.to_string())?)
-            } else if let BuiltInWordlist::OS = config.wordlist {
-                //Box::new(RawWordSelector::from_path(PathBuf::from(OS_WORDLIST_PATH))?)
-                let rws_result = RawWordSelector::from_path(PathBuf::from(OS_WORDLIST_PATH));
-                match rws_result {
+                let word_selector_result = RawWordSelector::from_path(PathBuf::from(wordlist_path));
+                match word_selector_result {
                     Ok(rws) => Box::new(rws),
-                    Err(e) => {
-                        // convert to toipe error
-                        let mut te = ToipeError::from(e);
-                        let msg_addition = "Error reading the given word list: ".to_string();
-                        te.msg = msg_addition + &te.msg;
-                        return Err(te);
-                    }
+                    Err(e) => return Err(create_toipe_wordlist_error(e)),
+                }
+            } else if let Some(word_list) = config.wordlist.contents() {
+                let word_selector_result = RawWordSelector::from_string(word_list.to_string());
+                match word_selector_result {
+                    Ok(rws) => Box::new(rws),
+                    Err(e) => return Err(create_toipe_wordlist_error(e)),
+                }
+            } else if let BuiltInWordlist::OS = config.wordlist {
+                let word_selector_result = RawWordSelector::from_path(PathBuf::from(OS_WORDLIST_PATH));
+                match word_selector_result {
+                    Ok(rws) => Box::new(rws),
+                    Err(e) => return Err(create_toipe_wordlist_error(e)),
                 }
             } else {
                 // this should never happen!

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -45,7 +45,7 @@ pub struct ToipeError {
 
 impl ToipeError {
     /// Prefixes the message with a context
-    pub fn with_context(mut self: Self, context: &str) -> Self {
+    pub fn with_context(mut self, context: &str) -> Self {
         self.msg = context.to_owned() + &self.msg;
         self
     }

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -73,33 +73,26 @@ impl<'a> Toipe {
     /// Initializes the word selector.
     /// Also invokes [`Toipe::restart()`].
     pub fn new(config: ToipeConfig) -> Result<Self, ToipeError> {
-        let word_selector_error_prefix = "Error reading the given word list: ";
-        let word_selector: Box<dyn WordSelector> =
+        let word_selector: Result<Box<dyn WordSelector>, _> =
             if let Some(wordlist_path) = config.wordlist_file.clone() {
-                let word_selector_result = RawWordSelector::from_path(PathBuf::from(wordlist_path));
-                match word_selector_result {
-                    Ok(rws) => Box::new(rws),
-                    Err(e) => return Err(ToipeError::from(e).with_context(word_selector_error_prefix)),
-                }
+                RawWordSelector::from_path(PathBuf::from(wordlist_path))
+                    .map(|ws| Box::new(ws) as Box<dyn WordSelector>)
             } else if let Some(word_list) = config.wordlist.contents() {
-                let word_selector_result = RawWordSelector::from_string(word_list.to_string());
-                match word_selector_result {
-                    Ok(rws) => Box::new(rws),
-                    Err(e) => return Err(ToipeError::from(e).with_context(word_selector_error_prefix)),
-                }
+                RawWordSelector::from_string(word_list.to_string())
+                    .map(|ws| Box::new(ws) as Box<dyn WordSelector>)
             } else if let BuiltInWordlist::OS = config.wordlist {
-                let word_selector_result = RawWordSelector::from_path(PathBuf::from(OS_WORDLIST_PATH));
-                match word_selector_result {
-                    Ok(rws) => Box::new(rws),
-                    Err(e) => return Err(ToipeError::from(e).with_context(word_selector_error_prefix)),
-                }
+                RawWordSelector::from_path(PathBuf::from(OS_WORDLIST_PATH))
+                    .map(|ws| Box::new(ws) as Box<dyn WordSelector>)
             } else {
                 // this should never happen!
                 // TODO: somehow enforce this at compile time?
-                return Err(ToipeError {
-                    msg: "Undefined word list or path.".to_string(),
-                });
+                Err(std::io::Error::new(
+                    std::io::ErrorKind::Other,
+                    "Undefined word list or path.",
+                ))
             };
+        let word_selector = word_selector
+            .map_err(|e| ToipeError::from(e).with_context("Error reading the given word list: "))?;
 
         let mut toipe = Toipe {
             tui: ToipeTui::new(),

